catroid/src/main/java/org/catrobat/catroid/content/bricks/FormulaBrick.java

```diff
@@ -170,4 +170,12 @@ public void storeDataForBackPack(Sprite sprite) {
 	public List<BackPackedVariableData> getBackPackedVariableData() {
 		return backPackedVariableData;
 	}
+
+	protected String getStringFromFloatForBrickView(float value) {
```

Comment:
I didn't really review your PR, but this is the first thing I noticed.
You're casting ints to floats a few times, e.g.
```java
getStringFromFloatForBrickView((float) BrickValues.ADD_ITEM_TO_USERLIST)
```
just to use this method. Instead, just overload it :wink:

However, when I type `1.0` into a formula and press ok, it will show `1` in the brick. I don't know where this transformation happens, but maybe there is some code you can reuse?

Reply:
hey,
I'm always (6 times) casting double to float for the prototype views. I think it's better to keep those six casts instead of overloading the method. 
If you want I can still overload this method but it would do exactly the same.

Your mentioned transformation is done in generateExternNumberStringForToken which I also refactored. In this method I get a String and cut the zeros at the end.

Reply:
> Your mentioned transformation is done in generateExternNumberStringForToken which I also refactored. In this method I get a String and cut the zeros at the end.

Great!
For consistency, it would be great if the prototype and real brick share the same logic to format a number/number string.

Reply:
I'll change the function for the prototypeview accordingly since this is the required return type.
Thx for the feedback

Reply:
forgot to write
made the functions consistent last week

Reply:
Thanks for the update! :smirk_cat:

k, for now let's take one step back and just focus on fixing the ticket. We can improve it later :wink: 

Would it work if we replace every wrong `String.valueOf` with `Utils.getNumberStringForBricks` and change corresponding doubles in BrickValues to ints or if that doesn't work, floats, without casting?

Reply:
Would it work if we replace every wrong String.valueOf with Utils.getNumberStringForBricks and change corresponding doubles in BrickValues to ints or if that doesn't work, floats, without casting?

Since all those values are only used for the Category lists in CategoryBricksFactory it should be possible to do that yes.

May I ask what's the problem with the current changes? They definitely fix the ticket and the refactored methods should also be better than before, I'm assuming you don't "like" the regex? (maybe slow or something I don't know?)

Reply:
> May I ask what's the problem with the current changes?…

Of course :+1: 
I'm not sure if there are tests that fail but there should be. You changed the behavior for `NUMBER` on `generateExternStringFromToken` ([pull/2710/files#diff-1acf93c2](https://github.com/Catrobat/Catroid/pull/2710/files#diff-1acf93c21492c193eec8057e7b6a7be5L289)). It should return a localized number (with coma or dot as decimal separator).

However, displaying numbers is currently inconsistent anyways. It will always show a `.` for prototypes or a localized separator for bricks. Fixing that will take some time as well and will introduce more changes. There also need to be tests.

This ticket is rather old. By quickly fixing it the way it is done in all other bricks, code quality and behavior will kinda stay the same. So we can finally close it and concentrate on improving later on.
___
I don't know the code @84n4n4 commented on, but:
```java
public static String getNumberStringForBricks(Context context, float value) {
	Formula formula = new Formula(value);
	return formula.getDisplayString(context);
}
```
should do a nice job to keep the displayed strings consistent. Otherwise you'll need to open up the Formula Editor which definitely won't be fun. But maybe I'm missing something here…?
Anyway, not important for this Pull Request. I also hope that the prototype view is complete removed in GSoC'18. I don't know why it is needed anyway.

Reply:
i've added the localized number thing so the behaviour should be the same as before. also tried your suggestion, but it does exact the opposite of what should happen. The String which gets returned from the getDisplayString is then always with ".0" . I can just get rid of it if u want me to. But the current code should do the same.

Reply:
Ah, yes. I meant `getTrimmedFormulaString`, not ~`getDisplayString`~. But this will be temporary as well, I guess. Let's drop that idea completely.

You missed a few `String.valueOf`s in Bricks. Even though they are `int`s, they've to be fixed as well.
___
@cheidenreich @84n4n4 @thmq, I'll approve this PR once all `valueOf`s are fixed. Trimming (removing `.0+`) numbers is done or should be done in a few more places (ShowTextActor, FE Data, etc.). I'll create a ticket so someone creates a single implementation for that.